docs/architecture/app-class-and-template.md

```diff
@@ -0,0 +1,200 @@
+
+
+## Terminology
+- **ApplicationClass** represents soon-to-be Application that is managed by IntegrationSystem. 
+   
+- **ApplicationTemplate** represents soon-to-be Application with all details already provided (like API and Events) that 
+can be created by providing values for replacing placeholders.
+  
+## GraphQL API Changes
+
+1. Managing Integration Systems
+ 
+```
+type IntegrationSystem {
+    id: ID!
+    name: String!
+    description: String
+    auths: [SystemAuth!]!
+}
+
+input IntegrationSystemInput {
+    name: String!
+    description: String
+}
+
+type Mutation {
+    createIntegrationSystem(in: IntegrationSystemInput!): IntegrationSystem!
+    updateIntegrationSystem(id: ID!, in: IntegrationSystemInput!): IntegrationSystem!
+    deleteIntegrationSystem(id: ID!): IntegrationSystem! 
+    
+    generateClientCredentialsForIntegrationSystem(id: ID!): SystemAuth!
+    deleteSystemAuthForIntegrationSystem(id: ID!): SystemAuth!
+}
+
+type Query {
+    integrationSystems(first: Int = 100, after: PageCursor): IntegrationSystemPage!
+    integrationSystem(id: ID!): IntegrationSystem
+}
+```
+Removing Integration System can be performed only when:
+- there is no Application Class from this Integration System
+- there is no Application managed by the Integration System 
+
+   
+2. Managing Application Class
+
+```
+input ApplicationClassInput {
+    name: String!
+    description: String
+    global: Boolean! = true
+    documents: [Document!]
+    integrationSystemID: ID!
+    apiSpecs: [ApiSpecInput!]!
+    eventApiSpecs: [EventApiSpecInput!]!
+}
+
+input ApplicationInput {
+    ...
+    applicationClassName: String
+    ...
+}
+
+
+type Mutation {
+    createApplicationClass(in: ApplicationClassInput!): ApplicationClass!
+    updateApplicationClass(id: ID!, in: ApplicationClassInput): ApplicationClass!
+    deleteApplicationClass(id: ID!): ApplicationCLass!
+}
+
+type Query {
+    applicationClasses(first: Int = 100, after: PageCursor): ApplicationClassPage!
+    applicationClass(id: ID): ApplicationClass!
+}
+
+```
+`ApplicationClass.documents` is a place for documenting any manual steps required for managing Application of given class by 
+a Integration System defined by `integrationSystemID`. `apiSpecs` and `eventApiSpecs` serves as a preview of API 
+or EventAPI provided by Applications originated from the ApplicationClass.
+In the first iteration, ApplicationClass can be registered only globally (`ApplicationClass.global`), so it will be available for every tenant. 
+To create the Application of given class, user set field `applicationClassName` in `ApplicationInput`. This field is optional.
+`deleteApplicationClass` is possible only if there are no classes originated from this ApplicationClass. 
+
+3. Managing Application Template 
+
+```
+input ApplicationTemplateInput {
+    name:               String!
+    applicationInput:   ApplicationInput!
+    placeholders:       [PlaceholderDefinitionInput]
```

Comment:
Right now not in the scope. Nice feature for future. Add small section about feature called: "Placehiolders"

Reply:
I added a section called `Application Template Placeholders Implementation` with an explanation of why placeholder implementation seems to be easy. Why am I pushing toward this approach? 
First of all, in my opinion, it does not make sense to provide templates without possibility of customizing it - it will cause that we can create only one application from given template in the given tenant. 

We also considered an option about adding a new query that returns applicationInput for the given template.
It has at least 2 drawbacks:
- ApplictionInput cannot be used as a return value (Graphql limitation), because of that we have to introduce a counterpart for it. And that causes code duplication. What is even worse an ApplicationInput has many other inputs and they also have to be duplicated. 
- using this API can be cumbersome. At the beginning, you have to query for all fields from 
from ApplicationInput (this is how Graphql works), the response will be formatted in JSON format and then you have to manually convert it to Graphql again, here you can find a simplified example:

query:
```
  inputFromTemplate{
   Name
    Description
  }
}
```
response:
```
{
  "data": {
    "inputFromTemplate": {
      "Name": "name",
      "Description": "desc"
    }
  }
}
```
and mutation:
```
mutation {
  createApp(sth:{Name:"name", Description:"desc"}) {
    Name
    Description
  }
}
```